src/common/src/util/resource_util.rs

```diff
@@ -0,0 +1,295 @@
+                super::CgroupVersion::V2 =>  format!("{}{}", super::DEFAULT_CGROUP_ROOT_HIERARCYHY, V2_MEMORY_LIMIT_HIERARCHY),
+            };
+            match super::read_integer_from_file_path(&limit_path){
+                Some(mem_limit_val) => return mem_limit_val,
+                None => return get_system_memory(),
+            }
+        }
+
+        // Returns the memory used in a container if running in a container else returns the system memory used.
+        pub fn get_memory_used_in_container(cgroup_version: super::CgroupVersion) -> usize{
+            if let super::CgroupVersion::V2 = cgroup_version{
+                if !super::is_controller_activated(super::Controller::Memory){
+                    return get_system_memory_used();
+                }
+            }
+            let usage_path = match cgroup_version{
+                super::CgroupVersion::V1 => format!("{}{}", super::DEFAULT_CGROUP_ROOT_HIERARCYHY, V1_MEMORY_CURRENT_HIERARCHY),
+                super::CgroupVersion::V2 =>  format!("{}{}", super::DEFAULT_CGROUP_ROOT_HIERARCYHY, V2_MEMORY_CURRENT_HIERARCHY),
+            };
+            match super::read_integer_from_file_path(& usage_path){
+                Some(mem_usage_val) => return mem_usage_val,
+                None => return get_system_memory_used(),
+            }
+        }
+
+        // Returns total memory used, if running in container, will return total memory used in container that process runs in.
+        pub fn total_memory_used_bytes() -> usize{
+            if super::is_linux_machine() && super::is_running_in_container() {
+                match super::get_cgroup_version() {
+                    Some(cgroup_version) => { get_memory_used_in_container(cgroup_version)},
+                    None => get_system_memory_used(),
+                }
+            }else{
+                get_system_memory_used()
+            }
+        }
+
+        // Returns total memory available, if running in container, will return total memory limit in container that process runs in.
+        pub fn total_memory_available_bytes() -> usize {
+            if super::is_linux_machine() && super::is_running_in_container(){
+                match super::get_cgroup_version() {
+                    Some(cgroup_version) => { get_container_memory_limit(cgroup_version)},
+                    None => get_system_memory(),
+                }
+            }else{
+                get_system_memory()
+            }
+        }
+    }
+
+    pub mod cpu{
+        use std::fs;
+        use num_cpus;
+
+         // Default constant Cgroup paths and hierarchy.
+        const V1_CPU_QUOTA_HIERARCHY: &str = "/cpu/cpu.cfs_quota_us";
+        const V1_CPU_PERIOD_HIERARCHY: &str = "/cpu/cpu.cfs_period_us";
+        const V2_CPU_LIMIT_HIERARCHY: &str = "/cpu.max";
+
+        // Returns the total number of CPU available, will return cpu limit if running in container.
+        pub fn total_cpu_available()-> f32 {
```

Comment:
Have you compared the result returned from this function and the [std::thread::available_parallelism](https://doc.rust-lang.org/stable/std/thread/fn.available_parallelism.html) in different environments?

I remember the std function does a similar thing. If they are the same, I prefer to use the std function for better portability.

Reply:
Just tested and they indeed have the same estimation. Have rectified to use [std:🧵:available_parallelism](https://doc.rust-lang.org/stable/std/thread/fn.available_parallelism.html). Do note that a panic will be thrown if available_parallelism returns an error which means -> platform not supported or process not have the right permission, which is the expected behavior. Thanks.